base/notify-api/api-deployment.yaml

```diff
@@ -283,20 +283,29 @@ spec:
               memory: "900Mi"
           ports:
             - containerPort: 6011
+          startupProbe:
+            httpGet:
+              path: /_status?simple=true
+              port: 6011
+            initialDelaySeconds: 15
+            periodSeconds: 3
+            timeoutSeconds: 1
+            successThreshold: 1
+            failureThreshold: 6
```

Comment:
The startup probe will make its first check after 15 seconds, after which it will try 6 times every 3 seconds, tagging the startup as a definitive failure after 33 seconds (15 + 6 * 3). 

An interesting issue we have with the slower New Relic initialization is that the latter is only triggered after the first request hits the API. That means that while the start up probe might be successful after the first request and then let the readiness and live probes to handle the rest, it might be misleading as New Relic would lock the app during its initialization.

Hence it is quite possible that we need to increase the `initialDelaySconds` of 15 seconds here.

Also note that new changes in gunicorn with regard to its timeout configuration will be as follows:

1. `gracefulTimeout` set to 25 seconds; in order to have the New Relic initialization roll out properly before gunicorn restarts the worker and in return, make the k8s pod reaches its own timeout.
2. `timeout` set to 30 seconds; this is the hard timeout for gunicorn to let its workers initialized.

Hence the 33 seconds failure for the k8s startupProbe is aligned with the hard timeout of gunicorn. That said, given the first request might be successful per reasons described above but make the rest of the application main thread locked, we might have to tweak this further.

Reply:
Nice description!

let’s try this